Add sppark-ntt

zlyber opened the pull request.

jizhuoran commented on transform.py:
This file should be deleted?

zlyber replied:
Sure, I have forgotten to delete it before submitting this PR.

jizhuoran commented on torch/nn/modules/ntt_zkp.py:
1. We need to rethink the name of the file and modular. How about we make a file name "zkp.py", and within this file, classes can be named as "ntt" rather than "NTT_ZKP".
2. I am wondering if the parameters can be sperated into ntt and intt, if so, should we make two class or not?

zlyber replied:
1. Sounds good at current stage, but I'm afraid this name can't be used forever if PNP will support more zkp protocols (e.g. zkSTARK). I think it would be better to use the protocol name "PLONK" or "zksnark". I will change the name of file and module in the next PR.
2. Do you mean we should generate intt and ntt parameters independently in sppark code or in python code? 
In the current code, we just create two instances, corresponding to intt and ntt.

jizhuoran commented on torch/nn/modules/__init__.py:
Please confirm we really need to modify this file.

zlyber replied:
There will be an error if we don't make this change:
```
Traceback (most recent call last):
  File "/home/zhiyuan/PNP/test_ntt.py", line 9, in <module>
    inttclass = nn.NTT_zkp(True, 0, domain_size, torch.BLS12_377_Fr_G1_Mont)
AttributeError: module 'torch.nn' has no attribute 'NTT_zkp'
```

jizhuoran replied:
emmm... To be honest, I have forgotten everything about my comments. But I think I put this comment because, at that time, I thought init.py does not need to be modified. Let me check this latter.

jizhuoran commented on torch/nn/functional.py:
I think we need to refactor this part. We only need on function, and use two parameters to distinguish ntt/intt coset-or-not. If you think it is hard, I can help.

zlyber replied:
Sorry I've forgotten why we need this part, L.O.L. I think we can make the distinction in [torch/nn/modules/ntt_zkp.py](https://github.com/PNP-team/PNP/pull/9/files/05a7ed6d4fa71701bb9143f5179d711956c31094#diff-709a5754612cbdcce9051d747c1e7944f5805ee4ffbb4b5bf11c1d63ea249b2a)

jizhuoran commented on torch/csrc/lazy/core/shape_inference.h:
We have confirmed that this file does not need to be changed.

jizhuoran commented on torch/csrc/lazy/core/shape_inference.cpp:
We have confirmed that this file does not need to be changed.

zlyber replied:
I see.

jizhuoran commented on test_ntt.py:
delete it or merge it to a.py. LOL, we really need a formal python file for testing... @wynneyin

zlyber replied:
I will merge it. BTW, doesn’t "a.py" seem too informal?

jizhuoran commented on elliptic_curve_model_with_tensors.png:
delete please

jizhuoran commented on build_log:
delete please

jizhuoran commented on elliptic_curve_model_with_tensors:
What is this file?

zlyber replied:
Maybe this is added by @wynneyin? I don't know either.

jizhuoran commented on aten/src/ATen/native/ts_native_functions.yaml:
No need to modify this file.

zlyber replied:
I see.

jizhuoran commented on aten/src/ATen/native/native_functions.yaml:
please clean this file. Meanwhile, see the below comment, we should use boolean parameter to distinguish ntt/intt coset

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/sppark-util/vec2d_t.hpp:
Please check if this file is needed.

zlyber replied:
It seems that we do not need this file anymore.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/sppark-ntt/nttkernels/algkernels.cuh:
the file name is not good. Meanwhile, please consider to re reorganize the dir names

zlyber replied:
The problem is that the bit_rev function is called by multiple files, including `aten/src/ATen/native/biginteger/cuda/sppark-ntt/nttkernels/gs_mixed_radix_wide.cuh`, `aten/src/ATen/native/biginteger/cuda/sppark-ntt/nttkernels/ct_mixed_radix_wide.cuh` and `aten/src/ATen/native/biginteger/cuda/sppark-ntt/nttkernels/kernels.cuh`. Maybe we can migrate those functions from `algkernels.cuh` to `kernels.cuh`.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/sppark-msm/sort.cuh:
MSM should be deleted in this branch?

zlyber replied:
I see.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/sppark-msm/pippenger.cuh:
MSM should be deleted in this branch?

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/sppark-msm/batch_addition.cuh:
MSM should be deleted in this branch?

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/sppark-msm.cu:
MSM should be deleted in this branch?

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ff/pasta.hpp:
I will handle this file.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ff/mont_t.cuh:
I will handle this file.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ff/mnt4753.hpp:
I will add this file via a seperated commit. But it lacks NTT parameters, right?

zlyber replied:
Yeah, next time when you want to support this field I will add those parameters. We can keep this file for now.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ff/goldilocks.hpp:
I will handle this file.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ff/gl64_t.cuh:
I will handle this file.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ff/bls12-381.hpp:
I will handle this file.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ff/bls12-381-fp2.hpp:
I will handle this file.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ff/bls12-377.hpp:
I will handle this file.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ff/bls12-377-fp2.hpp:
I will handle this file.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ff/bb31_t.cuh:
I will handle this file.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ff/baby_bear.hpp:
Why this is changed?

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ff/alt_bn128.hpp:
I will handle this file.

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ec/xyzz_t.hpp:
Please remove MSM related file from this commit

jizhuoran commented on aten/src/ATen/native/biginteger/cuda/ec/jacobian_t.hpp:
Please remove MSM related file from this commit

jizhuoran commented on aten/src/ATen/native/biginteger/cpu/ff/mnt4753.hpp:
I will handle this file

jizhuoran commented on aten/src/ATen/CMakeLists.txt:
It seems that I add a redundant new line. LOL

jizhuoran commented on a.py:
Please remove this file?

zlyber replied:
We need a function to test the correctness of our ntt. This file can be reserved until @wynneyin merge it to the testbench.

jizhuoran reviewed (commented):
I have commited every file. Maybe there are more comments, but let's fix them first.

zlyber commented on aten/src/ATen/native/biginteger/cuda/sppark-util/thread_pool_t.hpp:
It seems that we do not need this file anymore.

zlyber commented on aten/src/ATen/native/biginteger/cuda/sppark-ntt/parameters/parameters.cuh:
got it.

zlyber commented on aten/src/ATen/native/biginteger/cuda/sppark-ntt/parameters/gen_twiddles.cuh:
LGTM.

zlyber commented on aten/src/ATen/native/biginteger/cuda/sppark-ntt/nttkernels/ntt.cuh:
I agree.

zlyber commented on aten/src/ATen/native/biginteger/cuda/sppark-ntt/nttkernels/algorithm.cuh:
This was added by me... awkward.

zlyber commented on aten/src/ATen/native/biginteger/cuda/sppark-ntt/ntt_config.h:
Sppark uses `MAX_LG_DOMAIN_SIZE`  and the iteration num to determine  current four-step ntt size.

zlyber commented:
I have push the latest code of `zzy-to-merge`, please review.
Here is a description of some major changes:
1. On https://github.com/PNP-team/PNP/pull/9#discussion_r1432916198, I created 4 classes for ntt, intt and coset in `torch/nn/modules/zksnark.py`.
2. On https://github.com/PNP-team/PNP/pull/9#discussion_r1432927693 and，I deleted the directory sppark-util and all files in it, and I removed the custom cudaStream in ntt code.
3. On https://github.com/PNP-team/PNP/pull/9#discussion_r1432933273, I add a additional tensor in the func `params_zkp_cuda` at `sppark-ntt.cu` and dispatch it to gpu to store those local parameters such as root of unities.
4. On https://github.com/PNP-team/PNP/pull/9#discussion_r1432939371, I merge `sppark-ntt/algkernels.cuh` to `sppark-ntt/kernels.cuh`.
5. Remove all code related to MSM.
6. We still need discussion on https://github.com/PNP-team/PNP/pull/9#discussion_r1432940845 and https://github.com/PNP-team/PNP/pull/9#discussion_r1432935172.
